Nomad kills off healthy allocations on deploy when new allocations fail

### Nomad version

```
$ nomad -v
Nomad v0.10.2 (0d2d6e3dc5a171c21f8f31fa117c8a765eb4fc02)
```

### Operating system and Environment details

Amazon Linux 2 running on AWS.

### Issue

When we are deploying using the Blue/Green canary deployment model we have 4 allocations running (healthy), and we try to boot up 4 new allocations and they fail, Nomad retries that, and somehow Nomad kills off the healthy and running allocations leaving us with no healthy allocations running:

<img width="1193" alt="screenie_1576449035_369936" src="https://user-images.githubusercontent.com/334273/70913444-ed880700-200d-11ea-8874-be93c0aeb8e7.png">

### Reproduction steps

Deploy a healthy job, after that deploy a failing job, observe what happens.

### Job file

```hcl
job "billetto-production-web" {
  datacenters = ["dc1"]
  type = "service"

  update {
    max_parallel = 4
    min_healthy_time = "10s"
    healthy_deadline = "3m"
    progress_deadline = "10m"
    auto_revert = false
    auto_promote = false
    canary = 4
  }

  migrate {
    max_parallel = 3
    health_check = "checks"
    min_healthy_time = "10s"
    healthy_deadline = "5m"
  }

  group "group" {
    count = 4

    restart {
      attempts = 0
      interval = "30m"
      delay = "15s"
      mode = "fail"
    }

    ephemeral_disk {
      size = 300
    }

    task "nginx" {
      driver = "docker"

      config {
        image = "billetto/billetto-nginx:42.42.42"

        port_map {
          nginx = 80
        }

        dns_servers = ["172.17.0.1"]
      }

      resources {
        cpu    = 200
        memory = 170
        network {
          mbits = 50
          port "nginx" {}
        }
      }

      service {
        name = "billetto-production-nginx"
        tags = []
        port = "nginx"

        check {
          name     = "billetto-production-web nginx healthcheck"
          type     = "http"
          protocol = "http"
          path     = "/debug/healthcheck"
          interval = "5s"
          timeout  = "3s"
        }
      }
    }

    task "rails" {
      driver = "docker"

      config {
        image = "billetto/billetto:42.42.42"

        command = "bin/bundle"
        args = ["exec", "unicorn", "-c", "config/unicorn.rb", "--no-default-middleware"]

        port_map {
          web = 3000
        }

        dns_servers = ["172.17.0.1"]
      }

      resources {
        cpu    = 750
        memory = 1650
        network {
          mbits = 50
          port "web" {}
        }
      }

      service {
        name = "billetto-production-web"
        tags = []
        port = "web"

        check {
          name     = "billetto-production-web rails healthcheck"
          type     = "http"
          protocol = "http"
          path     = "/debug/healthcheck"
          interval = "5s"
          timeout  = "3s"
        }
      }
    }
  }
}
```

The node that runs the allocation `e8410fc6-0ff9-b52c-50ec-20b7de942ad5` writes this about it:

```
2020-01-09T14:00:42.789Z [ERROR] client.alloc_runner.task_runner.task_hook.stats_hook: failed to start stats collection for task with unrecoverable error: alloc_id=e8410fc6-0ff9-b52c-50ec-20b7de942ad5 task=rails error="container stopped"
2020-01-09T14:00:42.791Z [DEBUG] client.driver_mgr.docker.docker_logger: plugin process exited: driver=docker path=/usr/local/bin/nomad pid=10926
2020-01-09T14:00:42.791Z [DEBUG] client.driver_mgr.docker.docker_logger: plugin exited: driver=docker
2020-01-09T14:00:42.809Z [DEBUG] client.driver_mgr.docker: image id reference count decremented: driver=docker image_id=sha256:53afe179bd1fb99f1e87c61f5cfb013775542eb86ac07c28ac34bef3e494e7e7 references=20
2020-01-09T14:00:42.812Z [INFO]  client.gc: marking allocation for GC: alloc_id=e8410fc6-0ff9-b52c-50ec-20b7de942ad5
2020-01-09T14:00:42.817Z [DEBUG] client.alloc_runner.task_runner.task_hook.logmon: plugin process exited: alloc_id=e8410fc6-0ff9-b52c-50ec-20b7de942ad5 task=rails path=/usr/local/bin/nomad pid=10792
2020-01-09T14:00:42.817Z [DEBUG] client.alloc_runner.task_runner.task_hook.logmon: plugin exited: alloc_id=e8410fc6-0ff9-b52c-50ec-20b7de942ad5 task=rails
2020-01-09T14:00:42.817Z [DEBUG] client.alloc_runner.task_runner: task run loop exiting: alloc_id=e8410fc6-0ff9-b52c-50ec-20b7de942ad5 task=rails
```

Full node log here: 
[node-38adf12bfiltered.log](https://github.com/hashicorp/nomad/files/4040934/node-38adf12bfiltered.log)

Full leader log here: 
[leaderfiltered.log](https://github.com/hashicorp/nomad/files/4040939/leaderfiltered.log)

--- a/nomad/state/state_store.go
+++ b/nomad/state/state_store.go
@@ -3569,6 +3569,7 @@ func (s *StateStore) updateDeploymentWithAlloc(index uint64, alloc, existing *st
 		alloc.DeploymentStatus.ModifyIndex = index
 	}
 
+
 	// Create a copy of the deployment object
 	deploymentCopy := deployment.Copy()
 	deploymentCopy.ModifyIndex = index
@@ -3577,6 +3578,21 @@ func (s *StateStore) updateDeploymentWithAlloc(index uint64, alloc, existing *st
 	state.PlacedAllocs += placed
 	state.HealthyAllocs += healthy
 	state.UnhealthyAllocs += unhealthy
+
+	// Ensure PlacedCanaries accurately reflects the alloc canary status
+	if alloc.DeploymentStatus != nil {
+		if alloc.DeploymentStatus.Canary {
+			found := false
+			for _, canary := range state.PlacedCanaries {
+				if alloc.ID == canary {
+					found = true
+				}
+			}
+			if !found {
+				state.PlacedCanaries = append(state.PlacedCanaries, alloc.ID)
+			}
+		}
+	}
 
 	// Update the progress deadline
 	if pd := state.ProgressDeadline; pd != 0 {

--- a/nomad/state/state_store.go
+++ b/nomad/state/state_store.go
@@ -3569,7 +3569,6 @@ func (s *StateStore) updateDeploymentWithAlloc(index uint64, alloc, existing *st
 		alloc.DeploymentStatus.ModifyIndex = index
 	}
 
-
 	// Create a copy of the deployment object
 	deploymentCopy := deployment.Copy()
 	deploymentCopy.ModifyIndex = index
@@ -3580,17 +3579,16 @@ func (s *StateStore) updateDeploymentWithAlloc(index uint64, alloc, existing *st
 	state.UnhealthyAllocs += unhealthy
 
 	// Ensure PlacedCanaries accurately reflects the alloc canary status
-	if alloc.DeploymentStatus != nil {
-		if alloc.DeploymentStatus.Canary {
-			found := false
-			for _, canary := range state.PlacedCanaries {
-				if alloc.ID == canary {
-					found = true
-				}
+	if alloc.DeploymentStatus != nil && alloc.DeploymentStatus.Canary {
+		found := false
+		for _, canary := range state.PlacedCanaries {
+			if alloc.ID == canary {
+				found = true
+				break
 			}
-			if !found {
-				state.PlacedCanaries = append(state.PlacedCanaries, alloc.ID)
-			}
+		}
+		if !found {
+			state.PlacedCanaries = append(state.PlacedCanaries, alloc.ID)
 		}
 	}
 

--- a/nomad/state/state_store_test.go
+++ b/nomad/state/state_store_test.go
@@ -6382,6 +6382,109 @@ func TestStateStore_UpsertDeploymentAllocHealth_BadAlloc_Nonexistent(t *testing.
 	}
 }
 
+// Test that a deployments PlacedCanaries is properly updated
+func TestStateStore_UpsertDeploymentAlloc_Canaries(t *testing.T) {
+	t.Parallel()
+
+	state := testStateStore(t)
+
+	// Create a deployment
+	d1 := mock.Deployment()
+	require.NoError(t, state.UpsertDeployment(2, d1))
+
+	// Create a Job
+	job := mock.Job()
+	require.NoError(t, state.UpsertJob(3, job))
+
+	// Create alloc with canary status
+	a := mock.Alloc()
+	a.JobID = job.ID
+	a.DeploymentID = d1.ID
+	a.DeploymentStatus = &structs.AllocDeploymentStatus{
+		Healthy: helper.BoolToPtr(false),
+		Canary:  true,
+	}
+	require.NoError(t, state.UpsertAllocs(4, []*structs.Allocation{a}))
+
+	// Pull the deployment from state
+	ws := memdb.NewWatchSet()
+	deploy, err := state.DeploymentByID(ws, d1.ID)
+	require.NoError(t, err)
+
+	// Ensure that PlacedCanaries is accurate
+	require.Equal(t, 1, len(deploy.TaskGroups[job.TaskGroups[0].Name].PlacedCanaries))
+
+	// Create alloc without canary status
+	b := mock.Alloc()
+	b.JobID = job.ID
+	b.DeploymentID = d1.ID
+	b.DeploymentStatus = &structs.AllocDeploymentStatus{
+		Healthy: helper.BoolToPtr(false),
+		Canary:  false,
+	}
+	require.NoError(t, state.UpsertAllocs(4, []*structs.Allocation{b}))
+
+	// Pull the deployment from state
+	ws = memdb.NewWatchSet()
+	deploy, err = state.DeploymentByID(ws, d1.ID)
+	require.NoError(t, err)
+
+	// Ensure that PlacedCanaries is accurate
+	require.Equal(t, 1, len(deploy.TaskGroups[job.TaskGroups[0].Name].PlacedCanaries))
+
+	// Create a second deployment
+	d2 := mock.Deployment()
+	require.NoError(t, state.UpsertDeployment(5, d2))
+
+	c := mock.Alloc()
+	c.JobID = job.ID
+	c.DeploymentID = d2.ID
+	c.DeploymentStatus = &structs.AllocDeploymentStatus{
+		Healthy: helper.BoolToPtr(false),
+		Canary:  true,
+	}
+	require.NoError(t, state.UpsertAllocs(6, []*structs.Allocation{c}))
+
+	ws = memdb.NewWatchSet()
+	deploy2, err := state.DeploymentByID(ws, d2.ID)
+	require.NoError(t, err)
+
+	// Ensure that PlacedCanaries is accurate
+	require.Equal(t, 1, len(deploy2.TaskGroups[job.TaskGroups[0].Name].PlacedCanaries))
+}
+
+func TestStateStore_UpsertDeploymentAlloc_NoCanaries(t *testing.T) {
+	t.Parallel()
+
+	state := testStateStore(t)
+
+	// Create a deployment
+	d1 := mock.Deployment()
+	require.NoError(t, state.UpsertDeployment(2, d1))
+
+	// Create a Job
+	job := mock.Job()
+	require.NoError(t, state.UpsertJob(3, job))
+
+	// Create alloc with canary status
+	a := mock.Alloc()
+	a.JobID = job.ID
+	a.DeploymentID = d1.ID
+	a.DeploymentStatus = &structs.AllocDeploymentStatus{
+		Healthy: helper.BoolToPtr(true),
+		Canary:  false,
+	}
+	require.NoError(t, state.UpsertAllocs(4, []*structs.Allocation{a}))
+
+	// Pull the deployment from state
+	ws := memdb.NewWatchSet()
+	deploy, err := state.DeploymentByID(ws, d1.ID)
+	require.NoError(t, err)
+
+	// Ensure that PlacedCanaries is accurate
+	require.Equal(t, 0, len(deploy.TaskGroups[job.TaskGroups[0].Name].PlacedCanaries))
+}
+
 // Test that allocation health can't be set for an alloc with mismatched
 // deployment ids
 func TestStateStore_UpsertDeploymentAllocHealth_BadAlloc_MismatchDeployment(t *testing.T) {

--- a/scheduler/generic_sched.go
+++ b/scheduler/generic_sched.go
@@ -529,10 +529,6 @@ func (s *GenericScheduler) computePlacements(destructive, place []placementResul
 				// If we are placing a canary and we found a match, add the canary
 				// to the deployment state object and mark it as a canary.
 				if missing.Canary() && s.deployment != nil {
-					if state, ok := s.deployment.TaskGroups[tg.Name]; ok {
-						state.PlacedCanaries = append(state.PlacedCanaries, alloc.ID)
-					}
-
 					alloc.DeploymentStatus = &structs.AllocDeploymentStatus{
 						Canary: true,
 					}

--- a/scheduler/generic_sched_test.go
+++ b/scheduler/generic_sched_test.go
@@ -2029,17 +2029,28 @@ func TestServiceSched_JobModify_Canaries(t *testing.T) {
 	if plan.Deployment == nil {
 		t.Fatalf("bad: %#v", plan)
 	}
-	state, ok := plan.Deployment.TaskGroups[job.TaskGroups[0].Name]
-	if !ok {
-		t.Fatalf("bad: %#v", plan)
-	}
-	if state.DesiredTotal != 10 && state.DesiredCanaries != desiredUpdates {
-		t.Fatalf("bad: %#v", state)
-	}
+
+	// Ensure local state was not altered in scheduler
+	staleState, ok := plan.Deployment.TaskGroups[job.TaskGroups[0].Name]
+	require.True(t, ok)
+
+	require.Equal(t, 0, len(staleState.PlacedCanaries))
+
+	ws := memdb.NewWatchSet()
+
+	// Grab the latest state
+	deploy, err := h.State.DeploymentByID(ws, plan.Deployment.ID)
+	require.NoError(t, err)
+
+	state, ok := deploy.TaskGroups[job.TaskGroups[0].Name]
+	require.True(t, ok)
+
+	require.Equal(t, 10, state.DesiredTotal)
+	require.Equal(t, state.DesiredCanaries, desiredUpdates)
 
 	// Assert the canaries were added to the placed list
 	if len(state.PlacedCanaries) != desiredUpdates {
-		t.Fatalf("bad: %#v", state)
+		assert.Fail(t, "expected PlacedCanaries to equal desiredUpdates", state)
 	}
 }
 

